src/Orleans.Transactions/Abstractions/ITransactionalStateStorage.cs

```diff
@@ -0,0 +1,66 @@
+﻿using Orleans.Concurrency;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Orleans.Transactions.Abstractions
+{
+    public interface ITransactionalStateStorage<TState>
+        where TState : class, new()
+    {
+        Task<TransactionalStorageLoadResponse<TState>> Load(string stateName);
+
+        Task<string> Persist(
+            string stateName,
+            string expectedETag,
+            string metadata,
+            List<PendingTransactionState<TState>> statesToPrepare);
+
+        Task<string> Confirm(
+            string stateName,
+            string expectedETag,
+            string metadata,
+            string transactionIdToCommit);
+    }
+
+    [Serializable]
+    [Immutable]
+    public class PendingTransactionState<TState>
+        where TState : class, new()
+    {
+        public PendingTransactionState(string transactionId, long sequenceId, TState state)
+        {
+            this.TransactionId = transactionId;
+            this.SequenceId = sequenceId;
+            this.State = state;
+        }
+
+        public string TransactionId { get; }
+
+        public long SequenceId { get; }
+
+        public TState State { get; }
+    }
+
+    [Serializable]
+    [Immutable]
+    public class TransactionalStorageLoadResponse<TState>
+        where TState : class, new()
+    {
+        public TransactionalStorageLoadResponse(string etag, TState committedState, string metadata, List<PendingTransactionState<TState>> pendingStates)
```

Comment:
Not a big deal, but this ctor could expose `pendingStates` as an `IReadOnlyList` like the corresponding property.